### ISSUE TITLE
Guard: Constrained generic type to class/struct prevent any unconstraint use.

### ISSUE THREAD

**Issue by olivier-spinelli** [CommunityToolkit/dotnet#43]
## Overview

Being able to use struct instead of class for performance is more and more common. Hoewever, the Guard.IsNotNull cannot be used for such case. However, removing the constraints gives the exact same results.

Below is a concrete example:

```csharp
/// <summary>
/// Adapts a comparer and a value to a comparable.
/// To be used with <see cref="MemoryExtensions"/> binary search span extension methods.
/// </summary>
public readonly struct ComparerComparable<T, TComparer> : IComparable<T>
    where TComparer : IComparer<T>
{
    private readonly T _value;
    private readonly TComparer _comparer;

    public ComparerComparable( T value, TComparer comparer )
    {
        Guard.IsNotNull( value, nameof( value ) );
        Guard.IsNotNull( comparer, nameof( comparer ) );
        _value = value;
        _comparer = comparer;
    }

    [MethodImpl( MethodImplOptions.AggressiveInlining )]
    public int CompareTo( T? other ) => _comparer.Compare( _value, other );
}
```

This doesn't compile.
CS0452	The type 'T' must be a reference type in order to use it as parameter 'T' in the generic type or method 'Guard.IsNotNull<T>(T?, string)'

## Expected behavior

We should be able to use IsNotNull (and others) on unconstrained types.

## Environment
NuGet Package(s): 7.1.2 (latest)

## Additional context

I tested my hypothesis here: https://sharplab.io/#gist:32b79a5f6d137a29671f7a72c6083922 to be sure...
Basically, absolutely nothing changed, except that the following stupid code can be written:
```csharp
    // => 3 bytes. Stupid code leads to 0 overhead.
    public int GuardNCVal( int x ) {
        IsNotNullNC(x, nameof(x) );
        return x.GetHashCode();
    }
```
As the comment says, this is a no-op (check this on the SharpLab gist).

I will happily provide a PR for this if the idea is accepted (and there is no suble issues I missed).

**Comment by YegorStepanov** [comment]
`Null` and `default` struct are different concepts but under the hood they both equal to `00...` memory.

You should add a method like this:
```C#
static void IsNotNullOrDefault<T>(T? value, string name)
{
    if(Equals(value, default(T)))
    {
        throw new ArgumentException(name);
    }
}
```

I'm not tested it.

**Comment by olivier-spinelli** [comment]
@YegorStepanov  I'm afraid you missed the point. The goal here is to be able to use the Guard (and not redefining yet another helper) on unconstrained types rather than on value type and on reference type (Roslyn is just clever enough to do the job).

**Comment by YegorStepanov** [comment]
@olivier-spinelli I'm afraid of this too. 

I have read you want call one Guard method for class/struct in generic. 😄 I apologize for my drowsiness.

**Comment by Sergio0694** [comment]
This is a reasonable request, we should do this. My only concern is about binary breaking changes, I'll need to look into whether the `class` constraint might cause issue, given that that is encoded in IL metadata. If that's fine, then this looks good to me 😄

**Comment by olivier-spinelli** [comment]
Hi!
Finally, I decided to use a slightly different approach by using the new CallerArgumentExpressionAttribute (and it rocks!).
This is here if you want to have a look: https://github.com/Invenietis/CK-Core/tree/develop/CK.Core/Throw#a-simpler-approach-in-net6

**Comment by Sergio0694** [comment]
`[CallerArgumentExpression]` is already merged into `main` (#29), will be in the next preview 😉

**Comment by olivier-spinelli** [comment]
I misses this one!
But still: by using it I've decided to use one and only one `CheckArgument` method (actually there's a little bit more methods but that's the idea).
Thank youy for your excellent job by the way ! (and happy new year!)
